### ISSUE TITLE
Evaluation work stream

### ISSUE THREAD

**Issue by asuzukosi** [asuzukosi/pegasus#18]
**Work stream tracker.** Split into ordered issues; this one carries the selection and the rationale.

Pegasus has no benchmark harness, no task set, no scoring, and no baseline against any other harness. The README's claim to be *"the best agentic runtime harness for using open source models"* is unmeasured.

### Use Harbor — do not write a runner

[Harbor](https://harborframework.com/) is what deepagents runs on (`harbor[langsmith]>=0.20.0`). It provides sandbox environments (Docker, Modal, Daytona, E2B), automatic test execution and verification, reward scoring from 0.0 to 1.0, and **ATIF** trajectory logging. That is the runner and scorer already built, plus sandboxing and a standard trajectory format we would otherwise invent. We write an adapter, not a framework.

### The stream, in order

- [ ] **#44** — Drive pegasus non-interactively: `--json` and a Harbor adapter
- [ ] **#45** — Baseline comparison against mini-swe-agent
- [ ] **#46** — Public benchmarks and the open-model axis

#45 before #46 deliberately: the head-to-head is the cheapest number and the only one that speaks to the README's claim. Public benchmarks are for comparability with the field; the baseline is for knowing whether pegasus is worth using.

### Benchmark selection

| benchmark | measures | where |
|---|---|---|
| mini-swe-agent head-to-head | does pegasus beat the minimum | #45 |
| Terminal-Bench 2.0 | agent operating a computer via the terminal | #46 |
| SWE-bench Verified | real GitHub issues, patch correctness | #46 |
| τ²-bench | tool-agent-user interaction, multi-turn | #46 |
| the eight shipped models on one task set | the open-model claim | #46 |

**Excluded.** ARC-AGI-1/2 are static grid puzzles — the harness *does* influence the score through test-time compute (program synthesis, sampling, verification against training pairs), but scoring well needs a purpose-built loop that would not transfer to terminal work. ARC-AGI-3 is genuinely agentic but frame-based, so it mostly tests vision plumbing, and four of eight shipped models are text-only. Revisit if #20 lands — the REPL runtime is the right substrate for program synthesis. cua-bench waits for desktop control.

### Scoring convention

Taken from deepagents verbatim:

> **pass@K** — fraction of tasks that passed at least once within K rollouts.
> **avg@K** — passing trials (capped at K per task) ÷ expected trials (tasks × K); **missing rollouts count as failures, so a partial run can't inflate the score.**

That last clause is the one to enforce: infrastructure flakiness must not read as a good result.

### What this stream gates

- **#33** — per-model prompt variants cannot be chosen without it
- **#20** — the RLM decision is explicitly "decided on eval numbers, not taste"
- **#42** — refinement must be measured before it is automated

Depends on #14. There is no point measuring a harness whose compaction silently fails.

**Comment by asuzukosi** [comment]
## Which evals, ARC-AGI, and Harbor

### Use Harbor — do not write a runner

[Harbor](https://harborframework.com/) is the eval framework the field has settled on, and it is what deepagents uses (`harbor[langsmith]>=0.20.0` in `libs/evals/pyproject.toml`). From their CONTRIBUTING:

> Harbor is an evaluation framework that simplifies running agents on challenging benchmarks. It provides:
> - **Sandbox environments** (Docker, Modal, Daytona, E2B, etc.)
> - **Automatic test execution** and verification
> - **Reward scoring** (0.0 - 1.0 based on test pass rate)
> - **Trajectory logging** in ATIF format (Agent Trajectory Interchange Format)

That is the entire `evals/runners/` and `score.py` sketch in the issue body, already built, plus sandboxing and a standard trajectory format we would otherwise invent. Writing our own would be reimplementing a solved problem, which is exactly what we said not to do.

The integration point is an agent adapter — deepagents ships one at `deepagents_harbor/langgraph_project` and drives it with:

```bash
uv run harbor run --agent langgraph --dataset terminal-bench/terminal-bench-2 --model "$MODEL" -n 1 --env docker
```

**Revised scope for this issue: write a pegasus Harbor adapter, not an eval framework.** The adapter needs pegasus to be drivable non-interactively, which is the `--json` half of the editor-integration issue — worth pulling forward.

### Which benchmarks

| benchmark | what it measures | verdict |
|---|---|---|
| **Terminal-Bench 2.0** | 90+ tasks across software engineering, biology, security, gaming — agent operating a computer via the terminal | **Primary.** Harbor-native, closest to what pegasus is for, and gives a public comparable number |
| **SWE-bench Verified** | real GitHub issues, patch correctness | **Second.** The number everyone quotes; reuse mini-swe-agent's runner rather than writing one |
| **mini-swe-agent head-to-head** | same tasks, same models, different harness | **Do this first.** Cheapest, and the only thing that supports the README's claim |
| **τ³-bench** | multi-turn conversation against a simulated user | Later. Pegasus is not conversational-first |
| **cua-bench** | computer-use tasks, gym interface | Only once desktop control lands |

The open-model angle deserves its own axis and no existing benchmark covers it: **the same task set across all eight shipped models**, reporting where the harness compensates for weaker models and where it does not. That is the differentiated claim, and it is measurable with tasks we already have.

Take deepagents' scoring convention verbatim — pass@K and avg@K where **"missing rollouts count as failures, so a partial run can't inflate the score"**.

### ARC-AGI: mostly no, and here is the split

**ARC-AGI-1 and ARC-AGI-2 — not useful here.** Static grid-transformation puzzles. Pure model reasoning: no repo, no shell, no files, no long-horizon tool use. Running them through pegasus would measure the underlying model with pegasus adding overhead. It cannot answer "is pegasus a good harness" because the harness barely participates. If the goal is to publish a number about a *model*, use the model's API directly.

**ARC-AGI-3 — genuinely agentic, but the wrong fit for now.** [ARC-AGI-3-Agents](https://github.com/arcprize/ARC-AGI-3-Agents) is interactive game environments with a real loop — `FrameData`, actions, `levels_completed`, `win_levels` — and as of 0.9.3 supports local execution and custom environments rather than requiring the online API. So a harness does contribute. But:

- It is frame-based, so it tests vision plumbing above all. Four of the eight shipped models have `vision_support: False`.
- It measures perception and planning in novel games, not terminal or coding competence, which is what pegasus is for.
- It needs its own adapter and an ARC API key for the hosted path.

**Can we run it?** ARC-AGI-3: yes, by writing an agent in their harness that calls pegasus — a day or two of work. ARC-AGI-1/2: yes trivially, and the result would not mean anything about pegasus.

**Recommendation:** not in the first pass. Revisit ARC-AGI-3 after desktop computer use lands, when vision is a first-class path and the result would say something about pegasus rather than about the model.

**Comment by asuzukosi** [comment]
## Benchmark selection: locked

Going with **SWE-bench**, **Terminal-Bench 2.0**, and **τ²-bench**. All three run under Harbor, so the adapter is written once.

| benchmark | measures | why |
|---|---|---|
| **SWE-bench Verified** | real GitHub issues, patch correctness | the public number everyone quotes |
| **Terminal-Bench 2.0** | 90+ tasks, agent operating a computer via the terminal | closest to what pegasus is for; Harbor-native |
| **τ²-bench** | tool-agent-user interaction in real-world domains | multi-turn against a simulated user — the axis the other two miss |

τ²-bench is [sierra-research/tau2-bench](https://github.com/sierra-research/tau2-bench), live leaderboard at taubench.com. Note τ³ now exists (multimodal, voice, knowledge retrieval); τ² is the right target — text-only, stable, comparable numbers.

Dropped from the earlier list: ARC-AGI (see below), cua-bench (revisit when desktop control lands). Keeping the mini-swe-agent head-to-head as the first thing built, since it is cheapest and it is the only one that speaks to the README's claim.

## ATIF — what it is, and why we want it anyway

**Agent Trajectory Interchange Format.** From Harbor's docs:

> ATIF is a standardized, JSON-based specification for logging the complete interaction history of autonomous LLM agents. ATIF unifies the data requirements of conversational logs, action sequences, and replayable data structures, ensuring collected data is **immediately usable across debugging, visualization, Supervised Fine-Tuning (SFT), and Reinforcement Learning (RL) pipelines**.

It captures user messages, agent responses, tool executions, environment feedback, multi-turn structure, token usage and cost, and — relevant to us — **multi-agent systems with subagent delegation and hierarchical architectures**. Harbor ships Pydantic models (`harbor.models.trajectories`) and a validator. Claude Code, Codex, Gemini CLI, mini-SWE-agent and OpenHands all emit it.

Two reasons this matters beyond "Harbor wants it":

1. **Pegasus's `AgentEvent` stream already maps onto ATIF steps** — turn boundaries, tool calls, usage. Emitting ATIF is mostly a serialiser over an event stream that exists, and it is the same event stream the OTLP work consumes.
2. **It is training data.** ATIF is designed to feed SFT and RL pipelines. A harness that emits it produces usable post-training data as a side effect of running evals.

Making `AgentEvent` → ATIF the canonical trajectory output is the concrete first task of this issue, ahead of any benchmark.

## Correcting what I said about ARC-AGI

I overstated it. "The harness barely participates" is wrong, and worth correcting properly rather than leaving.

ARC-AGI-1/2 are static grid puzzles, so there is no repo, no shell, no file system — that part was right. But the harness is **not** irrelevant, because the scores that matter on ARC-AGI come from test-time compute, and test-time compute is a scaffold:

- **Program synthesis.** Strong approaches have the model write a candidate transformation program, run it against the task's training pairs, observe the failure, and revise. That is an agent loop with an execution tool and a verifier — exactly a harness.
- **Sampling and selection.** Generate many candidates, filter by which ones reproduce the training pairs, vote. The search policy is harness logic.
- **Verification is free and exact.** Every task ships input/output pairs, so a candidate can be checked without a judge. That is an unusually good fit for an agentic loop.

Published ARC-AGI results vary by large margins for the *same base model* depending on the scaffold. So a harness absolutely influences ARC-AGI.

**What I should have said:** ARC-AGI is a poor fit for pegasus *as currently shaped*, not a benchmark where harnesses do not matter. Scoring well needs a purpose-built program-synthesis loop — generate, execute against training pairs, revise, sample-and-vote — which pegasus does not have and which would not transfer to terminal or coding work. Building it would be building a second, specialised agent that happens to share a repo.

That is a scope judgement, not a claim about the benchmark. If ARC-AGI becomes a goal in its own right it is a legitimate and interesting target, and the RLM REPL mode is actually the right substrate for it — write a candidate program, run it, check it, iterate is precisely what that runtime is for. Worth revisiting after that decision, not before.
